python/perspective/perspective/table/table.py

```diff
@@ -303,6 +304,21 @@ def update(self, data, port_id=0):
                 data = "_" + data
 
         if _is_arrow:
+            if not _is_csv:
```

Comment:
This block incurs ~7 extra copies of the data. Even if optimized, I don't think this is the right location or mechanism (arrow->pandas->arrow) for such a change.

Reply:
I concur - though this codepath is already guilty of over-copying in at least a dozen other places.

I think this can be done similary in the data accessor code which creates [the column's null mask when reading numpy arrays](https://github.com/finos/perspective/blob/master/python/perspective/perspective/core/data/np.py#L21) without additional copying.

Reply:
@zxy994 Thanks for the update but this won't work either -
* Can't add `pyarrow` as a dependency - this is a _huge_ library impacting decision not to be made casually as a side effect of a bug fix.
* Still does multiple copy+iteration, but now to `pylist` which as _far worse_.
* .. now also stacks on a conversion from bytes->Arrow->bytes again.

More importantly, now that I've actually had time to review this PR unfortunately this approach just isn't going to work. In your example, you're not actually forcing update of `null` values - you're passing out-of-range values (`nan` for floats and `string` for _everything_ else). I do not think the issue reported can be fixed in Python, pretty sure it is due to how Arrow null values are [instantiated in `arrow_loader.cpp`](https://github.com/finos/perspective/blob/master/cpp/perspective/src/cpp/arrow_loader.cpp#LL617C6-L617C6).  IIRC this method sets the wrong `STATUS_INVALID` flag for updates - this is treated as an ignore, it needs to be `STATUS_CLEAR` for _update_ operations only (which may involve passing down a flag from all methods which call this).

Reply:
Understood, thanks! @texodus

Reply:
@texodus I have updated pr accordingly. I left out the formatting which was breaking =>_Error:   1) [perspective-cli-desktop-chrome] › ../../packages/perspective-cli/test/js/superstore.spec.js:15:5 › CLI › Tests something_

I looked into your earlier correction, and it's been interesting understanding the grander scale of things. I wanted to share my findings, and incur further correction/understanding where applies.

1. Binary and string datatypes are treated differently from eg a dictionary where both data has a None value. This part of code is called for an update using binary data=> https://github.com/finos/perspective/blob/master/cpp/perspective/src/cpp/arrow_loader.cpp#L320-L336  as compared to https://github.com/finos/perspective/blob/master/cpp/perspective/src/cpp/arrow_loader.cpp#L479-L483 .
2. Based on the above, I also find that using a binary data for update => https://github.com/finos/perspective/blob/master/cpp/perspective/src/cpp/arrow_loader.cpp#L603-L604 it reports a null count of zero even with a None value in data. This makes it bypass the set_valid method =>https://github.com/finos/perspective/blob/master/cpp/perspective/src/cpp/arrow_loader.cpp#L603-L604 which is called if update is done with a dictionary.
3. Albeit, I admit I am still looking into finding the link in using STATUS_CLEAR for update operations.